tip-distributor/src/lib.rs

```diff
@@ -455,59 +455,59 @@ pub fn derive_tip_distribution_account_address(
+    let mut blockhash = rpc_client
+        .get_latest_blockhash()
+        .await
+        .expect("fetch latest blockhash");
 
     let start = Instant::now();
-    while !transactions_to_send.is_empty() && start.elapsed() < max_send_duration {
-        info!("sending {} transactions", transactions_to_send.len());
-
-        for (signature, tx) in &transactions_to_send {
-            match rpc_client.send_transaction(tx).await {
-                Ok(send_tx_sig) => {
-                    info!("send transaction: {:?}", send_tx_sig);
-                }
-                Err(e) => {
-                    error!("error sending transaction {:?} error: {:?}", signature, e);
-                }
-            }
+    while start.elapsed() < max_retry_duration {
+        if start.elapsed() > Duration::from_secs(60) {
+            blockhash = rpc_client
+                .get_latest_blockhash()
+                .await
+                .expect("fetch latest blockhash");
         }
 
-        sleep(Duration::from_secs(10)).await;
+        transactions
+            .iter_mut()
+            .for_each(|tx| tx.sign(&[signer], blockhash));
+        let mut signatures_to_transactions = transactions
+            .clone()
+            .into_iter()
+            .map(|tx| (tx.signatures[0], tx))
+            .collect::<HashMap<Signature, Transaction>>();
 
-        let mut signatures_confirmed: Vec<Signature> = Vec::new();
-        for signature in transactions_to_send.keys() {
-            match rpc_client.confirm_transaction(signature).await {
-                Ok(true) => {
-                    info!("confirmed signature: {:?}", signature);
-                    signatures_confirmed.push(*signature);
-                }
-                Ok(false) => {
-                    info!("didn't confirmed signature: {:?}", signature);
-                }
-                Err(e) => {
-                    error!(
-                        "error confirming signature: {:?}, signature: {:?}",
-                        signature, e
-                    );
-                }
-            }
-        }
+        let futs = transactions.iter().map(|tx| async {
+            let result = rpc_client.send_and_confirm_transaction(tx).await;
+            (tx.signatures[0], result)
+        });
 
-        info!("confirmed {} signatures", signatures_confirmed.len());
-        for sig in signatures_confirmed {
-            transactions_to_send.remove(&sig);
-        }
+        errors = futures::future::join_all(futs)
+            .await
+            .into_iter()
+            .filter(|(_sig, result)| result.is_err())
+            .map(|(sig, result)| {
+                let e = result.err().unwrap();
+                warn!("error sending transaction: [error={e}, signature={sig}]");
+                (sig, e)
+            })
+            .collect::<HashMap<_, _>>();
+
+        transactions = errors
+            .iter()
+            .map(|(sig, _)| signatures_to_transactions.remove(sig).unwrap())
```

Comment:
you wanna remove the succesful ones, not the errored ones

Reply:
we wanna retry the failed ones

Reply:
can move the hashmap outside the loop then we'd rm only the succeeded ones and clone the errored ones for retry

Reply:
the commit that your commenting on wasn't using the `signatures_to_transactions` as a retry map and was getting initialized on each iteration, new code is different